### PR TITLE
Checkpoint ups for pretraining + example seq2seq generate

### PR DESCRIPTION
**Pull request by dpressel**
- adds a sample program for using pretrained paired seq2seq to do generation
- make the checkpoint type explicit in naming
- allow reloading from a checkpoint of step or epoch type
- suppport legacy checkpoint reloading with `--reload_tt (step|epoch)` option

### CODE REVIEW

File: api-examples/pretrain-transformer-lm.py [hunk 1]
```diff
@@ -526,10 +527,25 @@ def train():
     start_epoch = 0
     if args.restart_from:
         model.load_state_dict(torch.load(args.restart_from))
```
Comment (wenshuoliu) [inline api-examples/pretrain-transformer-lm.py]: Speaking of legacy ckpts, when i load the ones created by v1 and converted to v2, this line causes error because those ckpts don't have the overall layernorm. We probably want `strict=False` and a warning if there's mismatch of keys.

File: api-examples/pretrain-lm-streaming.py [hunk 2]
```diff
@@ -209,10 +210,24 @@ def train():
...
+        vec = args.restart_from.split("-")
+
+        if args.restart_tt:
+            tick_type = args.restart_tt
+        else:
+            tick_type = vec[-2]
+        step_num = int(vec[-1].split(".")[0])
+        if tick_type == 'epoch':
+            start_epoch = step_num - 1
+            global_step = (start_epoch + 1) * steps_per_epoch
```
Comment (wenshuoliu) [inline api-examples/pretrain-lm-streaming.py]: In all the cases, `steps_per_epoch` are calculated as `number_of_train_samples // args.batch_size`, but i think it should be `number_of_train_samples // (args.batch_size*num_gpus)` since each worker only needs to run this many steps to finish a epoch. I cut my release and tested that `int(os.environ.get("WORLD_SIZE", 1))` does give the num_gpus in our cluster environment. I still think we should fix this so the scheduler works as expected.

File: api-examples/pretrain-transformer-lm.py [hunk 3]
```diff
@@ -381,6 +381,7 @@ def train():
     parser.add_argument("--weight_decay", type=float, default=0.0, help="Weight decay")
     parser.add_argument("--epochs", type=int, default=20, help="Num training epochs")
     parser.add_argument("--restart_from", type=str, help="Option allows you to restart from a previous checkpoint")
+    parser.add_argument("--restart_tt", type=str, help="Optional param for legacy checkpoints (step|epoch)")
     parser.add_argument("--warmup_steps", type=int, default=1000, help="Num warmup steps")
     parser.add_argument("--mlm", type=str2bool, default=False, help="Use Masked Language Model (MLM) objective")
     parser.add_argument('--rpr_k', help='Relative attention positional sizes pass 0 if you dont want relative attention',
```
Comment (wenshuoliu) [inline api-examples/pretrain-transformer-lm.py]: Off the topic, but we still need lines like `if len(args.rpr_k) == 0 or args.rpr_k[0] < 1: args.rpr_k = None` in this file, or it'll complain.

File: api-examples/pretrain-transformer-lm.py (hunk 1 above)
Comment (dpressel) [reply api-examples/pretrain-transformer-lm.py]: Lets handle this in a separate PR

File: api-examples/pretrain-lm-streaming.py (hunk 2 above)
Comment (dpressel) [reply api-examples/pretrain-lm-streaming.py]: Can this be resolved?

File: api-examples/pretrain-transformer-lm.py (hunk 3 above)
Comment (dpressel) [reply api-examples/pretrain-transformer-lm.py]: Right, but I didnt remove that, missing the issue, can you clarify?

Comment (wenshuoliu) [reply api-examples/pretrain-transformer-lm.py]: It is in `pretrain-paired.py` and `pretrain-lm-streaming.py`, but it's missing in `pretrain-transformer-lm.py`.

Comment (dpressel) [reply api-examples/pretrain-transformer-lm.py]: fixed
